### PR TITLE
feat (hubpool): Add protocol fee capture method

### PR DESCRIPTION
**Pull request by chrismaree**
we need the ability for the protocol to capture fees. This PR adds this and associated unit tests.

### CODE REVIEW

File: test/HubPool.Fees.ts [hunk 1]
```diff
@@ -1,108 +0,0 @@
-import { toWei, toBNWei, SignerWithAddress, seedWallet, expect, Contract, ethers } from "./utils";
```
Comment (chrismaree) [inline test/HubPool.Fees.ts:1]: this file was a dup of the HubPool `HubPool.LiquidityProvisionFees.ts`

File: test/chain-adapters/Arbitrum_Adapter.ts [hunk 2]
```diff
@@ -76,12 +76,13 @@ describe("Arbitrum Chain Adapter", function () {
     // Create an action that will send an L1->L2 tokens transfer and bundle. For this, create a relayer repayment bundle
     // and check that at it's finalization the L2 bridge contracts are called as expected.
     const { leafs, tree, tokensSendToL2 } = await constructSingleChainTree(dai, 1, arbitrumChainId);
-    await hubPool.connect(dataWorker).initiateRelayerRefund([3117], 1, tree.getHexRoot(), consts.mockTreeRoot);
```
Comment (chrismaree) [inline test/chain-adapters/Arbitrum_Adapter.ts:79]: there was a merge issue from the previous PR. this is unrelated to the main change.

File: contracts/HubPool.sol [hunk 3]
```diff
@@ -73,7 +73,15 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
 
     // Interest rate payment that scales the amount of pending fees per second paid to LPs. 0.0000015e18 will pay out
     // the full amount of fees entitled to LPs in ~ 7.72 days, just over the standard L2 7 day liveness.
-    uint256 lpFeeRatePerSecond = 1500000000000;
+    uint256 public lpFeeRatePerSecond = 1500000000000;
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply contracts/HubPool.sol:76]: can we make this uint64 to match the other pcts?

File: contracts/HubPool.sol [hunk 4]
```diff
@@ -73,7 +73,15 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
...
+
+    mapping(address => uint256) public unclaimedAccumulatedProtocolFees;
+
+    // Address that captures protocol fees. Accumulated protocol fees can be claimed by this address.
+    address public protocolFeeCaptureAddress;
+
+    // Percentage of lpFees that are captured by the protocol and claimable by the protocolFeeCaptureAddress.
+    uint256 public protocolFeeCapturePct;
```
Comment (nicholaspai) [inline contracts/HubPool.sol:84]: Same, should change to uint64

File: contracts/HubPool.sol [hunk 5]
```diff
@@ -144,12 +158,23 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
...
         lpTokenFactory = _lpTokenFactory;
         finder = _finder;
+        protocolFeeCaptureAddress = owner();
     }
 
     /*************************************************
      *                ADMIN FUNCTIONS                *
      *************************************************/
 
+    function setProtocolFeeCaptureAddress(address newProtocolFeeCaptureAddress) public onlyOwner {
```
Comment (nicholaspai) [inline contracts/HubPool.sol:168]: Should this function logic be collapsed with the following one?

File: contracts/HubPool.sol [hunk 6]
```diff
@@ -405,6 +433,14 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
         delete refundRequest;
     }
 
+    function claimProtocolFeesCaptured(address l1Token) public nonReentrant {
+        if (unclaimedAccumulatedProtocolFees[l1Token] > 0) {
```
Comment (nicholaspai) [inline contracts/HubPool.sol:437]: do we need this check..? whats the harm of sending 0, the caller will jsut pay gas

File: test/chain-adapters/Arbitrum_Adapter.ts [hunk 7]
```diff
@@ -76,12 +76,13 @@ describe("Arbitrum Chain Adapter", function () {
...
+    await hubPool
+      .connect(dataWorker)
+      .initiateRelayerRefund([3117], 1, tree.getHexRoot(), consts.mockTreeRoot, consts.mockTreeRoot);
```
Comment (nicholaspai) [inline test/chain-adapters/Arbitrum_Adapter.ts:81]:
```suggestion
      .initiateRelayerRefund([3117], 1, tree.getHexRoot(), consts.mockDestinationDistributionRoot, consts.mockSlowRelayFulfillmentRoot);
```

File: test/chain-adapters/Arbitrum_Adapter.ts [hunk 8]
```diff
@@ -99,7 +100,7 @@ describe("Arbitrum Chain Adapter", function () {
       owner.address,
       consts.sampleL2Gas,
       consts.sampleL2GasPrice,
-      mockSpoke.interface.encodeFunctionData("initializeRelayerRefund", [consts.mockTreeRoot])
+      mockSpoke.interface.encodeFunctionData("initializeRelayerRefund", [consts.mockTreeRoot, consts.mockTreeRoot])
```
Comment (nicholaspai) [inline test/chain-adapters/Arbitrum_Adapter.ts:103]:
```suggestion
      mockSpoke.interface.encodeFunctionData("initializeRelayerRefund", [consts.mockDestinationDistributionRoot, consts.mockSlowRelayFulfillmentRoot])
```

**Review by nicholaspai — APPROVED** [review]

File: contracts/HubPool.sol (hunk 5 above)
Comment (chrismaree) [reply contracts/HubPool.sol:168]: ye, good point. can do how we do the `setBond` method.

File: contracts/HubPool.sol (hunk 6 above)
Comment (chrismaree) [reply contracts/HubPool.sol:437]: it's cheaper. it's just for gas optimisation.

I agree with you though that in this context we can remove this. In the context of the fee allocation I think it's better to keep this to remove gas usage.

File: contracts/HubPool.sol (hunk 4 above)
Comment (mrice32) [reply contracts/HubPool.sol:84]: Can we also put all the uint64 variables next to one another to decrease the cost of reading?

File: contracts/HubPool.sol [hunk 9]
```diff
@@ -144,12 +156,22 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
...
     /*************************************************
      *                ADMIN FUNCTIONS                *
      *************************************************/
 
+    function setProtocolFeeCapture(address newProtocolFeeCaptureAddress, uint256 newProtocolFeeCapturePct)
+        public
+        onlyOwner
+    {
+        protocolFeeCaptureAddress = newProtocolFeeCaptureAddress;
+        protocolFeeCapturePct = newProtocolFeeCapturePct;
```
Comment (mrice32) [inline contracts/HubPool.sol:171]: Should we check that this is <= 1e18?

File: contracts/HubPool.sol [hunk 10]
```diff
@@ -592,6 +619,22 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
...
+        // Calculate the fraction of bundledLpFees that are allocated to the protocol and to the LPs.
+        uint256 protocolFeesCaptured = (bundleLpFees * protocolFeeCapturePct) / 1e18;
+        uint256 lpFeesCaptured = bundleLpFees - protocolFeesCaptured;
+
+        // Assign any LP fees included into the bundle to the pooled token. These LP fees are tracked in the
+        // undistributedLpFees and within the utilizedReserves. undistributedLpFees is gradually decrease
+        // over the smear duration to give the LPs their rewards over a period of time. Adding to utilizedReserves
+        // acts to track these rewards after the smear duration. See _exchangeRateCurrent for more details.
+        pooledTokens[l1Token].undistributedLpFees += lpFeesCaptured;
+        pooledTokens[l1Token].utilizedReserves += int256(lpFeesCaptured);
```
Comment (mrice32) [inline contracts/HubPool.sol:632]: Should we wrap this in `if (lpFeesCaptured > 0)`?

**Review by mrice32 — APPROVED** [review]
LGTM! A few minor nits
